### PR TITLE
66 backendml make dimensional reduction methods generic

### PR DESCRIPTION
**Pull request by Varlog22**
Transferred the functionality of dimensionality reduction from the qaqc_main-model to the parent model(shared_model.py). This can now be used from the qaqc_main-model by sending embeddings to the function "dim_red" which returns a Pandas Dataframe with reduced embeddings.

### CODE REVIEW

**Review by WTANKW — APPROVED** [review]
Tested running the code, looks good, well done!

**Review by jacomoel — APPROVED** [review]
Tested to run the changes, works like a charm nice job!

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py [hunk 1]
```diff
@@ -71,32 +71,8 @@ def first_run(self, df: pd.DataFrame, dim: str | None) -> pd.DataFrame:
         id_df = pd.DataFrame(data=d)
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/sub_models/qaqc_main_model.py:71]: I know that this is not added to this commit, but I think you could clean up the concatinations by writing this:

```python
ids = [self.get_id(question) for question in questions]
df["id"] = pd.Series(ids)
```

If you add this you don't need the variable `d` and you can also remove the line `combined_df = pd.concat([combined_df, id_df], axis=1)`.

I think this is fitting, given that your code cleans up this part, even though I should have caught this in a previous review.

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py [hunk 2]
```diff
@@ -71,32 +71,8 @@ def first_run(self, df: pd.DataFrame, dim: str | None) -> pd.DataFrame:
...
-
-            case "UMAP":
-                umap_model = umap.UMAP(n_components=2, random_state=42, n_neighbors=15, min_dist=0.1, metric='euclidean')
-                reduced_embeddings_umap = umap_model.fit_transform(embeddings)
-                point_data_df = pd.DataFrame(reduced_embeddings_umap, columns=["x", "y"])
 
-            case _: 
-                raise Exception("Invalid dimension reduction method.")
-            
+        point_data_df = self.dim_red(embeddings=embeddings, dim=dim)
```
Comment (Leohemmingsson) [inline Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: I think this line would be even more fitting placed in the shared models  `run` function. That way no one can accidentally forget to add the function. 

I would probably place it on line 40 in `shared_model.py`

**Review by Leohemmingsson — CHANGES_REQUESTED** [review]
Good job, I like the solution. There are just two small things I think would make this a little better

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py (hunk 1 above)
Comment (Varlog22) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:71]: great catch, will do in my current branch for issue 69

File: Backend/src/ml_lib/sub_models/qaqc_main_model.py (hunk 2 above)
Comment (Varlog22) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: It's a good suggestion, however I believe that it can pose a problem with how we are handling the output from "latter_run()" at the moment. Will keep it in mind though

Comment (Leohemmingsson) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: I thought that both `first_run` and `latter_run` would return pandas data frames and that both would look similar with the only difference being how the nodes are clustered or placed. Is that not the case?

Comment (Varlog22) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: That is indeed the case, however I am not quite sure how one would perform dim_red() on embeddings that havent been made yet(as this is made in the functions called from run()). I guess you could perhaps manipulate the DataFrame extracted from the first_run() or latter_run() in order to perform the embedding and dimension reduction in run(). 

However,  it may not be optimal considering that we are using a highly specific sentence transformer model(to our dataset) to extract our embeddings. There may be some obvious solution that I am not seeing, but then you'll have to illuminate me.

Comment (Leohemmingsson) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: Ah, I see. Well I guess it depends on exactly how the `latter_run` is implemented, so it might be a good strategy to implement that first and then see if using `dim_red` in `run` still holds. We could cycle back to this when you guys are done with the first random forest implementation.

Comment (Varlog22) [reply Backend/src/ml_lib/sub_models/qaqc_main_model.py:75]: Sounds like a plan
